mail_sendgrid/tests/test_mail_sendgrid.py

```diff
@@ -0,0 +1,290 @@
+                                                  'sendgrid')
+
+        # Send mail
+        mail = self.create_email()
+        response = mail.send(mail)
+
+        self.assertTrue(response)
+        self.assertEqual(mock_sendgrid.called, True)
+        self.assertEqual(mail.state, 'sent')
+        mail_tracking = mail.tracking_email_ids
+        self.assertEqual(len(mail_tracking), 1)
+        self.assertFalse(mail_tracking.state)
+
+        # Test mail processed
+        self.event.update({
+            'event': 'processed',
+            'odoo_id': mail.message_id
+        })
+        response = self.env['mail.tracking.email'].event_process(
+            self.request, self.event, self.metadata)
+        self.assertEqual(response, 'OK')
+        self.assertEqual(mail_tracking.state, 'sent')
+
+        # Test mail delivered
+        self.event['event'] = 'delivered'
+        self.env['mail.tracking.email'].event_process(
+            self.request, self.event, self.metadata)
+        self.assertEqual(mail_tracking.state, 'delivered')
+        self.assertEqual(mail_tracking.recipient, self.recipient.email)
+        self.assertFalse(mail.opened)
+
+        # Test mail opened
+        self.event['event'] = 'open'
+        self.env['mail.tracking.email'].event_process(
+            self.request, self.event, self.metadata)
+        self.assertEqual(mail_tracking.state, 'opened')
+        self.assertTrue(mail.opened)
+
+        # Test click e-mail
+        self.event['event'] = 'click'
+        self.env['mail.tracking.email'].event_process(
+            self.request, self.event, self.metadata)
+        self.assertEqual(mail_tracking.state, 'opened')
+        self.assertEqual(mail.click_count, 1)
+
+        # Test events are linked to e-mail
+        mail.env.clear()
+        self.assertEqual(len(mail.tracking_event_ids), 4)
+
+    def test_controller(self):
+        """ Check the controller is working """
+        event_data = [self.event]
+        result = self.json_open(
+            '/mail/tracking/sendgrid/' + self.session.db,
+            json.dumps(event_data)
+        )
+        self.assertTrue(result)
+        # Invalid request
+        result = self.json_open(
+            '/mail/tracking/sendgrid/' + self.session.db,
+            "[{'invalid': True}]"
+        )
+        self.assertTrue(result)
+
+    @mock.patch(mock_template_api_client)
+    @mock.patch(mock_config_template)
+    def test_update_templates(self, m_config, m_sendgrid):
+        m_config.return_value = "ldkfjsOIWJRksfj"
+        m_sendgrid.return_value = FakeTemplateClient()
+        self.env['sendgrid.template'].update_templates()
+        template = self.env['sendgrid.template'].search([
+            ('remote_id', '=', 'fake_id')
+        ])
+        self.assertTrue(template)
+
+    def json_open(self, url, data=None, timeout=5):
+        if url.startswith('/'):
+            url = f"http://{HOST}:{PORT}{url}"
```

Comment:
You cannot use f'strings with python 3.5. Change this one with a regular string format.